language/move-prover/src/prelude.bpl

```diff
@@ -1131,6 +1111,40 @@ procedure {:inline 1} Signature_ed25519_threshold_verify(bitmap: Value, signatur
 
 // TODO: implement the below methods
 
-procedure {:inline 1} $LCS_to_bytes(ta: TypeValue, v: Reference) returns (res: Value) {
-    assert false; // $LCS_to_bytes not implemented
-}
+// ==================================================================================
+// Native LCS::serialize
+
+// native public fun serialize<MoveValue>(v: &MoveValue): vector<u8>;
+
+// Serialize is modeled as (almost) uninterpreted function, with an additional
+// axiom to say it's an injection.
+
+function $serialize(ta: TypeValue, v: Value): Value;
+
+// This says that $serialize respects isEquals (substitution property)
+// Without this, Boogie will get false positives where v1, v2 differ at invalid
+// indices.
+axiom (forall ta: TypeValue ::
+       (forall v1,v2: Value :: IsEqual(v1, v2) ==> IsEqual($serialize(ta, v1), $serialize(ta, v2))));
+
+
+// This says that serialize is an injection
+axiom (forall ta1, ta2: TypeValue ::
+       (forall v1, v2: Value :: IsEqual($serialize(ta1, v1), $serialize(ta2, v2))
+           ==> IsEqual(v1, v2) && (ta1 == ta2)));
+
+
+procedure $LCS_to_bytes(ta: TypeValue, r: Reference) returns (res: Value);
+ensures res == $serialize(ta, $Dereference($m, r));
+ensures $IsValidU8Vector(res);    // result is a legal vector of U8s.
+ensures $vlen(res) > 0;
+ensures is#Boolean($Dereference($m, r)) ==> $vlen(res) == 1;
```

Comment:
Why is this needed? Also, if we do it for bool, why not for others? (Could not understand what you mean with "wrong" below.)

Reply:
We know the lengths of some primitives.  I don't know if it matters, but my guess is that it has low cost for Boogie/Z3 and that it may help sometimes.  One scenario where it might make a difference is when concatenating serialized values.  If  x != z, you can conclude that x*y != z*w IF len(x) == len(z).  That may be important at some point (guid in libra_account concatenates some things and wants them to be unique, but I haven't tried it out to see if this is needed).

Reply:
I still don't understand why you have then commented out for the other primitives, and what you mean with "wrong". 

Also someone looking at this in the future, might not know why you added this, and gets confused somehow thinking this is a semantic requirement. If you keep it please add a comment why its there.

Reply:
Can we perhaps just leave this length out as we are right now have just a guess that it might be helpful, and it anyway doesn't work for anything else than bool? I find it highly confusing as it is now.